### ISSUE TITLE
previewer: tar.gz (and more) previewer

### ISSUE THREAD

**Issue by krzysztof** [zenodo/zenodo#557]
Add previewer for tarball (tar.gz), and perhaps some of the other popular archive format.

- .tar
- .tar.gz
- .tar.bz2
- .tbz
- .tbz2
- .7z

**Comment by shark-S** [comment]
I like to work on this issue. Can you explain a little more?

**Comment by krzysztof** [comment]
@shark-S You'll need to take a close look at the ZIP format previewer in the ``invenio-previewer`` extensions here: https://github.com/inveniosoftware/invenio-previewer/tree/master/invenio_previewer/extensions and see how it works in Zenodo (and also the whole invenio-previewer module).

Then create something similar to ZIP previewer for another format. I am not sure which of the ones above would be the easiest to start with, but I guess something like .tar or .tar.gz should allow you to look at the header of the archive without unpacking it.

**Comment by rajat1994** [comment]
@shark-S .... are you still working on this issue?.. If not, i would like to work on it.

**Comment by rajat1994** [comment]
@krzysztof .... do we have to create a whole zenodo previewer for this one like invenio previewer?

**Comment by krzysztof** [comment]
@rajat1994 Yes, it should be a new extension file like this: https://github.com/inveniosoftware/invenio-previewer/blob/master/invenio_previewer/extensions/zip.py

**Comment by krzysztof** [comment]
It should be a PR to Invenio module. Here's a guide on how to checkout Invenio module locally: http://zenodo.readthedocs.io/en/latest/development/guide.html

So test with Zenodo, but push all the code to invenio-previewer extensions

**Comment by rajat1994** [comment]
@rupak0557 ... i am not working on it right now. You can start working on it....If you need some help, do tell us.

**Comment by rupak0577** [comment]
Working on it

**Comment by dmey** [comment]
Any updated on this?

**Comment by bschilder** [comment]
Also curious if this has been implemented yet. Great idea!

**Comment by cuihaoleo** [comment]
> Then create something similar to ZIP previewer for another format. I am not sure which of the ones above would be the easiest to start with, but I guess something like .tar or .tar.gz should allow you to look at the header of the archive without unpacking it.

@krzysztof Unfortunately a tarfile (at least the GNU and POSIX variants) doesn't provide a nice mechanism to obtain the file list because you only know the file name when you reach the corresponding block ([ref](https://www.gnu.org/software/tar/manual/html_node/Standard.html)):
> Each file archived is represented by a header block which describes the file, followed by zero or more blocks which give the contents of the file.

This means that you would have to unpack everything (even in memory) to get the list.

zip/7z/rar support this better.

**Comment by ashwinvis** [comment]
@cuihaoleo You can list the files without unpacking and in my experience it is relatively fast. Quoting from the man pages.

`man tar`
```
       -t, --list
              List the contents of an archive.  Arguments are optional.   When
              given, they specify the names of the members to list.
```
`man bsdtar`
```
     -t      List archive contents to stdout.  The long option form is --list.
```

There is a Python method too. https://docs.python.org/3/library/tarfile.html#tarfile.TarFile.list

**Comment by cuihaoleo** [comment]
@ashwinvis Well, under the hood it still decompresses everything in the memory. I don't think there is any magic.

If the tarfile is not compressed at all, the tool can skip file content to minimize IO. If it's compressed, the tarfile is not seekable (unless using some extensions like pixz) and decompression must be done ([ref](https://en.wikipedia.org/wiki/Tar_(computing)#Random_access)).

Even worse, it has to be read all the way to the end to confirm that all files have been iterated, because: _At the end of the archive file there are two 512-byte blocks filled with binary zeros as an end-of-file marker._ ([ref](https://www.gnu.org/software/tar/manual/html_node/Standard.html)).

If you felt `tar -t` is faster (or even significantly faster) than `tar -x`, that's only because writing to disk is usually slow and `tar -t` needs not do that.

Here is an experiment if you want to test it yourself. If you using a standard Linux system (macOS may also work), you can create a stupidly highly compressed tar like this (compression could take a while):
```
$ truncate -s 16G file1
$ truncate -s 16G file2
$ tar -cvf- file1 file2 | xz -T0 --fast > test.tar.xz
```

I ended up compressing two 16G zero-filled files into a 8.1 MB tar.xz file. Then you can timing `tar -t` and in-memory decompression. It should take roughly the same time (actually in my case `tar -t` is even slower):

```
$ time tar tf test.tar.xz
file1
file2
tar tf test.tar.xz  65.08s user 35.64s system 150% cpu 1:07.15 total
$ time xzcat test.tar.xz > /dev/null
xzcat test.tar.xz > /dev/null  57.72s user 0.36s system 99% cpu 58.084 total
```

If the previewer function is indeed to be implemented, the best way I can think is to try decompress for example first hundreds MBs and extract whatever paths there. However, I'd look into the worst case CPU/memory usage to understand security implications.
